File: mobile/ios/PortaliaCore/Sources/PortaliaCore/SIP/SIPService.swift
import Foundation
import Combine

public enum SIPRegistrationState: Equatable {
    case notConfigured
    case registering
    case registered
    case failed(String)
}

public struct SIPCallInfo: Identifiable, Equatable {
    public let id = UUID()
    public let remoteDisplayName: String
    public let isIncoming: Bool
}

/// Interface que a UI (telas de Interfone) consome — não amarra a nenhuma
/// biblioteca SIP específica. `SIPServiceStub` é a única implementação hoje:
/// nenhum Asterisk real está conectado na V01 (mesma ressalva do backend,
/// ver docs/V01_SCOPE.md), então o app nunca finge estar registrado. Quando
/// houver domínio SIP real, uma implementação real (Linphone SDK + CallKit)
/// entra aqui atrás da mesma interface — nenhuma tela muda.
public protocol SIPService: AnyObject {
    var registrationState: CurrentValueSubject<SIPRegistrationState, Never> { get }
    var activeCall: CurrentValueSubject<SIPCallInfo?, Never> { get }

    func register(username: String, password: String, domain: String) async
    func unregister() async
    func placeCall(to destination: String, displayName: String) async
    func answerCall() async
    func endCall() async
    func declineCall() async
}

/// Implementação honesta: sem SIP_DOMAIN configurado no backend (ver
/// GET /resident/telephony/credentials `configured:false`), a tela de
/// Interfone nunca deve nem chamar `register` — mas se chamar, isto garante
/// que o estado nunca finge "registered" sem um domínio real.
public final class SIPServiceStub: SIPService {
    public let registrationState = CurrentValueSubject<SIPRegistrationState, Never>(.notConfigured)
    public let activeCall = CurrentValueSubject<SIPCallInfo?, Never>(nil)

    public init() {}

    public func register(username: String, password: String, domain: String) async {
        registrationState.send(.failed("Nenhum servidor Asterisk configurado nesta instalação."))
    }

    public func unregister() async {
        registrationState.send(.notConfigured)
    }

    public func placeCall(to destination: String, displayName: String) async {
        // Intencionalmente no-op: sem registro SIP real, não há nada para discar.
    }

    public func answerCall() async {}
    public func endCall() async { activeCall.send(nil) }
    public func declineCall() async { activeCall.send(nil) }
}

File: mobile/android/core/src/main/kotlin/com/portalia/core/sip/SipService.kt
package com.portalia.core.sip

import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow

sealed class SipRegistrationState {
    object NotConfigured : SipRegistrationState()
    object Registering : SipRegistrationState()
    object Registered : SipRegistrationState()
    data class Failed(val reason: String) : SipRegistrationState()
}

data class SipCallInfo(val remoteDisplayName: String, val isIncoming: Boolean)

/**
 * Interface que a UI (telas de Interfone) consome — não amarra a nenhuma
 * biblioteca SIP específica. `SipServiceStub` é a única implementação hoje:
 * nenhum Asterisk real está conectado na V01 (mesma ressalva do backend e
 * do lado iOS), então o app nunca finge estar registrado.
 */
interface SipService {
    val registrationState: StateFlow<SipRegistrationState>
    val activeCall: StateFlow<SipCallInfo?>

    suspend fun register(username: String, password: String, domain: String)
    suspend fun unregister()
    suspend fun placeCall(destination: String, displayName: String)
    suspend fun answerCall()
    suspend fun endCall()
    suspend fun declineCall()
}

class SipServiceStub : SipService {
    private val _registrationState = MutableStateFlow<SipRegistrationState>(SipRegistrationState.NotConfigured)
    override val registrationState: StateFlow<SipRegistrationState> = _registrationState

    private val _activeCall = MutableStateFlow<SipCallInfo?>(null)
    override val activeCall: StateFlow<SipCallInfo?> = _activeCall

    override suspend fun register(username: String, password: String, domain: String) {
        _registrationState.value = SipRegistrationState.Failed("Nenhum servidor Asterisk configurado nesta instalação.")
    }

    override suspend fun unregister() {
        _registrationState.value = SipRegistrationState.NotConfigured
    }

    override suspend fun placeCall(destination: String, displayName: String) {
        // Intencionalmente no-op: sem registro SIP real, não há nada para discar.
    }

    override suspend fun answerCall() {}
    override suspend fun endCall() { _activeCall.value = null }
    override suspend fun declineCall() { _activeCall.value = null }
}
